File: version4j/src/main/java/com/only4/algorithm/version4j/leetcode/Generate.java
package com.only4.algorithm.version4j.leetcode;

import java.util.ArrayList;
import java.util.List;

public class Generate {
    public List<List<Integer>> generate(int numRows) {
        // 参数校验
        if (numRows <= 0) return new ArrayList<>();

        // 创建杨辉三角结果列表
        List<List<Integer>> pascalTriangle = new ArrayList<>();

        // 为每一行创建列表并初始化为1（处理边界元素）
        for (int rowIndex = 0; rowIndex < numRows; rowIndex++) {
            List<Integer> currentRow = new ArrayList<>();
            for (int columnIndex = 0; columnIndex <= rowIndex; columnIndex++) {
                currentRow.add(1);  // 边界元素初始值都为1
            }
            pascalTriangle.add(currentRow);
        }

        // 从第3行(索引为2)开始填充中间元素
        for (int rowIndex = 2; rowIndex < numRows; rowIndex++) {
            for (int columnIndex = 1; columnIndex < rowIndex; columnIndex++) {
                // 当前元素 = 上一行左上方元素 + 上一行右上方元素
                int leftParentValue = pascalTriangle.get(rowIndex - 1).get(columnIndex - 1);
                int rightParentValue = pascalTriangle.get(rowIndex - 1).get(columnIndex);
                int currentValue = leftParentValue + rightParentValue;

                pascalTriangle.get(rowIndex).set(columnIndex, currentValue);
            }
        }

        return pascalTriangle;
    }
}

File: version4k/src/main/kotlin/com/only4/algorithm/version4k/leetcode/generate.kt
package com.only4.algorithm.version4k.leetcode

fun generate(numRows: Int): List<List<Int>> {
    // 参数校验
    if (numRows <= 0) return emptyList()

    // 创建杨辉三角结果列表，每行长度为rowIndex+1，初始值都为1
    val pascalTriangle = List(numRows) { rowIndex ->
        MutableList(rowIndex + 1) { 1 }
    }

    // 从第3行(索引为2)开始填充中间元素
    for (rowIndex in 2 until numRows) {
        for (columnIndex in 1 until rowIndex) {
            // 当前元素 = 上一行左上方元素 + 上一行右上方元素
            val leftParentValue = pascalTriangle[rowIndex - 1][columnIndex - 1]
            val rightParentValue = pascalTriangle[rowIndex - 1][columnIndex]
            val currentValue = leftParentValue + rightParentValue

            pascalTriangle[rowIndex][columnIndex] = currentValue
        }
    }

    return pascalTriangle
}
